src/metabase/models/user_parameter_value.clj

```diff
@@ -51,25 +51,46 @@
                 :dashboard_id dashboard-id
                 :parameter_id parameter-id)))
 
+(mu/defn batched-upsert!
+  "Delete param with nil value and upsert the rest."
+  [user-id         :- ms/PositiveInt
+   dashboard-id    :- ms/PositiveInt
+   param-id->value :- [:map-of :string :any]]
+  (let [to-delete (filter (comp nil? second) param-id->value)
+        to-upsert (filter (comp some? second) param-id->value)]
+    (t2/with-transaction [_conn]
+      (doseq [[param-id value] to-upsert]
+        (or (pos? (t2/update! :model/UserParameterValue {:user_id      user-id
```

Comment:
if we add a unique constraint on `parameter_id, dashboard_id, user_id` we can do this in one statement
```clojure
(t2/query {:insert-into :user_parameter_value
                  :values      (map (fn [[param-id value]]
                                      {:user_id      user-id
                                       :dashboard_id dashboard-id
                                       :parameter_id param-id
                                       :value        value})
                                    to-upsert)
                  :on-conflict  [:user_id :dashboard_id :parameter_id]
                  :do-update-set :value})
```

Reply:
Is there any reason not to add it?

Reply:
note: you'll have to write `merge into` for h2 and `insert ... update` for mysql though :)